### PR TITLE
Rswag Setup

### PR DESCRIPTION
**Pull request by srollins01**
Installed rswag gem and created rswag spec files as a starting point for API documentation. 
API documentation can be accessed at http://localhost:3000/api-docs.

### CODE REVIEW

File: spec/requests/api/v1/expertises_spec.rb [hunk 1]
```diff
@@ -0,0 +1,137 @@
+require 'swagger_helper'
+
+RSpec.describe 'api/v1/expertises', type: :request do
+
+  path '/api/v1/users/{user_id}/expertises' do
+    parameter name: 'user_id', in: :path, type: :string, description: 'user_id'
```
Comment (ayau8) [inline spec/requests/api/v1/expertises_spec.rb]: Suggest we update the description to e.g. "The ID of user that needs to be fetched" for better understanding, and global change to other 'user_id' accordingly

File: spec/requests/api/v1/expertises_spec.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    get('list expertises') do
```
Comment (ayau8) [inline spec/requests/api/v1/expertises_spec.rb]: Suggest we align description with spec tests, for example "list expertises" to "Returns the user's expertises" (capitalising the first letter of first word), and global change to other actions accordingly

File: spec/requests/api/v1/expertises_spec.rb [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      tags 'Expertises'
+      response(200, 'successful') do
```
Comment (ayau8) [inline spec/requests/api/v1/expertises_spec.rb]: Missing a 404 response

File: spec/requests/api/v1/expertises_spec.rb [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            }
+          }
+        end
+        run_test!
+      end
+    end
+
+    post('create expertise') do
+      tags 'Expertises'
+      response(200, 'successful') do
```
Comment (ayau8) [inline spec/requests/api/v1/expertises_spec.rb]: Based on latest master, response status should be 201 for CREATE

Also missing a 422 response

File: spec/requests/api/v1/expertises_spec.rb [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    end
+  end
+
+  path '/api/v1/users/{user_id}/expertises/{id}' do
+    parameter name: 'user_id', in: :path, type: :string, description: 'user_id'
+    parameter name: 'id', in: :path, type: :string, description: 'id'
+
+    get('show expertise') do
+      tags 'Expertises'
+      response(200, 'successful') do
```
Comment (ayau8) [inline spec/requests/api/v1/expertises_spec.rb]: Missing a 404 response

File: spec/requests/api/v1/expertises_spec.rb [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            'application/json' => {
+              example: JSON.parse(response.body, symbolize_names: true)
+            }
+          }
+        end
+        run_test!
+      end
+    end
+
+    patch('update expertise') do
```
Comment (ayau8) [inline spec/requests/api/v1/expertises_spec.rb]: We haven't included PATCH in our latest approved master, suggest we don't include this in our docs in the meantime

File: spec/requests/api/v1/expertises_spec.rb [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            }
+          }
+        end
+        run_test!
+      end
+    end
+
+    put('update expertise') do
+      tags 'Expertises'
+      response(200, 'successful') do
```
Comment (ayau8) [inline spec/requests/api/v1/expertises_spec.rb]: Missing a 422 response

File: spec/requests/api/v1/expertises_spec.rb [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            }
+          }
+        end
+        run_test!
+      end
+    end
+
+    delete('delete expertise') do
+      tags 'Expertises'
+      response(200, 'successful') do
```
Comment (ayau8) [inline spec/requests/api/v1/expertises_spec.rb]: Based on latest master, response status should be 204 for DELETE

Also missing a 404 response

File: spec/requests/api/v1/users_spec.rb [hunk 9]
```diff
@@ -0,0 +1,133 @@
+require 'swagger_helper'
+
+RSpec.describe 'api/v1/users', type: :request do
+
+  path '/api/v1/users' do
+
+    get('list users') do
```
Comment (ayau8) [inline spec/requests/api/v1/users_spec.rb]: Similar to Expertise's, suggest we align description with spec tests

File: spec/requests/api/v1/users_spec.rb [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      end
+    end
+
+    post('create user') do
+      tags 'Users'
+      response(200, 'successful') do
+        consumes 'application/json'
+        parameter name: :user, in: :body, schema: {
+          type: :object,
+          properties: {
```
Comment (ayau8) [inline spec/requests/api/v1/users_spec.rb]: Suggest we update the properties based on User's latest schema, and global changes to other User's actions accordingly

File: spec/requests/api/v1/users_spec.rb [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+      end
+    end
+  end
+
+  path '/api/v1/users/{id}' do
+    parameter name: 'id', in: :path, type: :string, description: 'id'
+
+    get('show user') do
+      tags 'Users'
+      response(200, 'successful') do
```
Comment (ayau8) [inline spec/requests/api/v1/users_spec.rb]: Missing a 404 response

File: spec/requests/api/v1/users_spec.rb [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+            'application/json' => {
+              example: JSON.parse(response.body, symbolize_names: true)
+            }
+          }
+        end
+        run_test!
+      end
+    end
+
+    patch('update user') do
```
Comment (ayau8) [inline spec/requests/api/v1/users_spec.rb]: Similar to Expertises, suggest we remove it if not included in spec test

File: spec/swagger_helper.rb [hunk 13]
```diff
@@ -18,16 +18,16 @@
...
       info: {
-        title: 'API V1',
+        title: 'Menternshift API V1',
         version: 'v1'
       },
       paths: {},
       servers: [
         {
-          url: 'https://{defaultHost}',
+          url: 'https://localhost:3000/',
```
Comment (ayau8) [inline spec/swagger_helper.rb]: Not sure if that's only my side, but "https" doesn't work for me when I try to play around the docs
It works if "http" is used though, hence suggest we double check on this

**Review by ayau8 — COMMENTED** [review]
Thanks Sarah for setting up the RSwag and relevant spec files

Understand the current version is a starting point and I've included comments regarding the actions' description, missing JSON response and server url for your consideration. Please update accordingly (including the yaml) as you see fit

Separately, note that we've included the User's doc (on top of the approved User#Show actions) - as we're in the process of approving the User's controller & spec testings, suggest we prioritise that approval, so we can update our User docs in one go based on approved version

File: spec/swagger_helper.rb (hunk 13 above)
Comment (srollins01) [reply spec/swagger_helper.rb]: Double-checked this and confirmed it should be http for localhost.

**Comment by srollins01** [comment]
@ayau8 Thanks for your review. I've made the relevant changes and also updated the expertise schema in the rswag docs. 
Will make any other necessary updates here once the user specs + controller have been approved.

**Comment by ayau8** [comment]
@srollins01 thanks for the update. Some more comments for consideration:

1) For User & Expertise's POST, suggest we update the parameter part similar to below, we have a "no parameters" in User's which might be confusing to reader

Screenshot 2024-03-19 at 4 52 10 PM

2) For schemas, not sure if that's only my side, but it seems that we need to expand the data type of each row if needed, can we present all the data types of each param once expanding the schema?

**Review by ayau8 — COMMENTED** [review]
@srollins01 thanks for the update. Some more comments for consideration:

For User & Expertise's POST, suggest we update the parameter part similar to below, we have a "no parameters" in User's which might be confusing to reader

Screenshot 2024-03-19 at 4 52 10 PM

For schemas, not sure if that's only my side, but it seems that we need to expand the data type of each row if needed, can we present all the data types of each param once expanding the schema?

**Comment by srollins01** [comment]
@ayau8 Thanks for having a look. 
For User & Expertise's POST I agree that the layout in the image you shared is easier to read. However, it looks like that is for an older version of OAS. Openapi 3.0.x does not support body params, but instead only supports path params like this: 
Screenshot 2024-03-20 114340

I suggest we stick with the most up-to-date version. More info here: https://swagger.io/docs/specification/describing-request-body/

Also, I've updated the UI config so all models should now have their data types expanded by default.

**Review by ayau8 — DISMISSED** [review]

**Review by ayau8 — COMMENTED** [review]
Thanks Sarah for the updates, no additional comments from my side. As we have had the User's controller and spec test approved and merged to master, can you please also help update API docs accordingly like the Expertise's?

**Review by ayau8 — DISMISSED** [review]

**Review by ayau8 — APPROVED** [review]

**Review by achpsident — APPROVED** [review]

**Pull request merged**